---
title: "GRN Inference from CLL time-course bulk RNA-seq"
author: "Malvina Marku, Hugo Chenel, Flavien Raynal"
output:
  pdf_document: default
  html_notebook: default
  html_document:
    df_print: paged
editor_options: 
  chunk_output_type: console
---


**1. Load the libraries and `dynGENIE3` functions:** 

```{r include=FALSE}
library(reshape2)
library(doRNG)
library(doParallel)
library(tidyverse)
library(DESeq2)
library(clusterSim)
library(dplyr)
library(ggplot2)
library(CINNA)
library(igraph)
library(qgraph)
library(ggVennDiagram)
library(gridExtra)

source("dynGENIE3.R")
```


**2. Load the data**
In our case, we have the time-course transcriptomics from 2 patients and 2 replicates per patient. the inference method will be applied in different ways:
\item patient 1 and 2 taken separately. The average between replicates will be calculated per patient.
\item ...

```{r include=FALSE}
Patient_1_rep_1_table <- read.expr.matrix("/mnt/SERVER-CRCT-STORAGE/CRCT21/Private/CLL_dec2020/RNAseqBulk_Eq21_Eq9/Gene_counts_tables/Gene_counts_Patient1_rep1_TPM.txt",form = "rows.are.genes")

Patient_1_rep_2_table <- read.expr.matrix("/mnt/SERVER-CRCT-STORAGE/CRCT21/Private/CLL_dec2020/RNAseqBulk_Eq21_Eq9/Gene_counts_tables/Gene_counts_Patient1_rep2_TPM.txt", form = "rows.are.genes")

Patient_2_rep_1_table <- read.expr.matrix("/mnt/SERVER-CRCT-STORAGE/CRCT21/Private/CLL_dec2020/RNAseqBulk_Eq21_Eq9/Gene_counts_tables/Gene_counts_Patient2_rep1_TPM.txt", form = "rows.are.genes")

Patient_2_rep_2_table <- read.expr.matrix("/mnt/SERVER-CRCT-STORAGE/CRCT21/Private/CLL_dec2020/RNAseqBulk_Eq21_Eq9/Gene_counts_tables/Gene_counts_Patient2_rep2_TPM.txt", form = "rows.are.genes")
```


```{r}
TS1 = read.expr.matrix("Data/time_series_1.txt", form="rows.are.samples")
TS2 = read.expr.matrix("Data/time_series_2.txt", form="rows.are.samples")
TS3 = read.expr.matrix("Data/time_series_3.txt", form="rows.are.samples")
```


**Re-format the data:**

_Firstly_, make sure there is no `NA, NaN, or Inf` value in the dataset: 

```{r include=FALSE}
#Identify the NA, NaN and Inf values
impute.mean <- function(x) replace(x, is.na(x) | is.nan(x) | is.infinite(x) | is.factor(x), mean(x, na.rm = TRUE))

Patient_1_rep_1_table <- apply(Patient_1_rep_1_table, 2, impute.mean)
sum( apply( Patient_1_rep_1_table, 2, function(.) sum(is.infinite(.))) )

Patient_1_rep_2_table <- apply(Patient_1_rep_2_table, 2, impute.mean)
sum( apply( Patient_1_rep_2_table, 2, function(.) sum(is.infinite(.))) )


```


```{r}
time.points = list(TS1[1,], TS2[1,], TS3[1,])
TS.data = list(TS1[2:nrow(TS1),], TS2[2:nrow(TS2),], TS3[2:nrow(TS3),])
```



_Secondly_, re-format the time-series:

```{r include=FALSE}
#Add a row in each of the datasets indicating the time point

Patient_1_time_points <- c(1:ncol(Patient_1_rep_1_table))
Patient_2_time_points <- c(1:ncol(Patient_2_rep_1_table))

#Remove the rows with 0 variance
Patient_1_rep_1_table <- rbind(Patient_1_time_points, Patient_1_rep_1_table)
Patient_1_rep_1_table <- Patient_1_rep_1_table[rowSums(Patient_1_rep_1_table[]) > 0,]

Patient_1_rep_2_table <- rbind(Patient_1_time_points, Patient_1_rep_2_table)
Patient_1_rep_2_table <- Patient_1_rep_2_table[rowSums(Patient_1_rep_2_table[]) > 0,]

Patient_2_rep_1_table <- rbind(Patient_2_time_points, Patient_2_rep_1_table)
Patient_2_rep_1_table <- Patient_2_rep_1_table[rowSums(Patient_2_rep_1_table[]) > 0,]

Patient_2_rep_2_table <- rbind(Patient_2_time_points, Patient_2_rep_2_table)
Patient_2_rep_2_table <- Patient_2_rep_2_table[rowSums(Patient_2_rep_2_table[]) > 0,]

# write.table(Patient_1_rep_1_table, file = "Patient_1_rep_1_table.txt")
# write.table(Patient_1_rep_2_table, file = "Patient_1_rep_2_table.txt")
# write.table(Patient_2_rep_1_table, file = "Patient_2_rep_1_table.txt")
# write.table(Patient_2_rep_2_table, file = "Patient_2_rep_2_table.txt")

# Patient_1_rep_1_table <- read.expr.matrix("Patient_1_rep_1_table.txt",form = "rows.are.genes")
# Patient_1_rep_2_table <- read.expr.matrix("Patient_1_rep_2_table.txt", form = "rows.are.genes")
# Patient_2_rep_1_table <- read.expr.matrix("Patient_2_rep_1_table.txt", form = "rows.are.genes")
# Patient_2_rep_2_table <- read.expr.matrix("Patient_2_rep_2_table.txt", form = "rows.are.genes")

### Normalize data in [0, 1] -- as in the example model
### Normalization method: n4 (https://rdrr.io/cran/clusterSim/man/data.Normalization.html)
#Patient_1_rep_1_unit = as.data.frame(data.Normalization(Patient_1_rep_1_table, type = "n4", normalization = "column" ))
#Patient_1_rep_2_unit = as.data.frame(data.Normalization(Patient_1_rep_2_table, type = "n4", normalization = "column" ))
#Patient_2_rep_1_unit = as.data.frame(data.Normalization(Patient_2_rep_1_table, type = "n4", normalization = "column" ))
#Patient_2_rep_2_unit = as.data.frame(data.Normalization(Patient_2_rep_2_table, type = "n4", normalization = "column" ))

#Find genes that appear in both df
common_genes_1 <- intersect(rownames(Patient_1_rep_1_table), rownames(Patient_1_rep_2_table))
common_genes_2 <- intersect(rownames(Patient_2_rep_1_table), rownames(Patient_2_rep_2_table))
common_genes <- intersect(common_genes_1, common_genes_2)

#Keep only the genes that appear in common_genes
Patient_1_rep_1_filter <- as.matrix(subset(Patient_1_rep_1_table, rownames(Patient_1_rep_1_table) %in% common_genes))
Patient_1_rep_2_filter <- as.matrix(subset(Patient_1_rep_2_table, rownames(Patient_1_rep_2_table) %in% common_genes))
Patient_2_rep_1_filter <- as.matrix(subset(Patient_2_rep_1_table, rownames(Patient_2_rep_1_table) %in% common_genes))
Patient_2_rep_2_filter <- as.matrix(subset(Patient_2_rep_2_table, rownames(Patient_2_rep_2_table) %in% common_genes))

#p1_rep1 = do.call(rbind.data.frame, time_series_data[1])
#p1_rep2 = do.call(rbind.data.frame, time_series_data[2])
#p2_rep1 = do.call(rbind.data.frame, time_series_data[3])
#p2_rep2 = do.call(rbind.data.frame, time_series_data[4])

time_series_data <- list(Patient_1_rep_1_filter[1:nrow(Patient_1_rep_1_filter),], Patient_1_rep_2_filter[1:nrow(Patient_1_rep_2_filter),], Patient_2_rep_1_filter[1:nrow(Patient_2_rep_1_filter),], Patient_2_rep_2_filter[1:nrow(Patient_2_rep_2_filter),])

df_time_series_data = as.data.frame(time_series_data)

sample_info = as.data.frame(str_split_fixed(colnames(df_time_series_data),pattern = "_", n = 3))
names(sample_info)[names(sample_info) == "V1"] <- "Day"
names(sample_info)[names(sample_info) == "V2"] <- "Patient"
names(sample_info)[names(sample_info) == "V3"] <- "Replicate"

#Create DESeq2 object
dds <- DESeqDataSetFromMatrix(countData = round(df_time_series_data), colData = sample_info, design = ~ Day + Patient + Replicate)
nrow(dds) #16432
#Remove rows having low counts
firstcondition <- rowSums(counts(dds) > 1) >= 2 #at least 2 samples with a count of 2 or higher
dds <- dds[firstcondition,]
nrow(dds) #12247

#Generate normalized counts
dds <- estimateSizeFactors(dds)
sizeFactors(dds)
normalized_counts <- counts(dds, normalized=TRUE)

#Remove all genes with zero variance (i.e constant expression)
variances = rowVars(normalized_counts)
range(variances)
normalized_counts = normalized_counts[!rowVars(normalized_counts) == 0 ,]
normalized_counts = as.data.frame(normalized_counts)

#List of matrices corresponding to the different patients/replicates 
df_conditions = sapply(c("p1_rep1", "p1_rep2", "p2_rep1", "p2_rep2"),
                function(x) normalized_counts[endsWith(names(normalized_counts),x)],
                simplify = FALSE)

df_conditions$p1_rep1 = as.matrix(df_conditions$p1_rep1)
df_conditions$p1_rep2 = as.matrix(df_conditions$p1_rep2)
df_conditions$p2_rep1 = as.matrix(df_conditions$p2_rep1)
df_conditions$p2_rep2 = as.matrix(df_conditions$p2_rep2)
normalized_counts_list <- list(df_conditions$p1_rep1[1:nrow(df_conditions$p1_rep1),], df_conditions$p1_rep2[1:nrow(df_conditions$p1_rep2),], df_conditions$p2_rep1[1:nrow(df_conditions$p2_rep1),], df_conditions$p2_rep2[1:nrow(df_conditions$p2_rep2),])


# Create the time points
time.points_1 <- c(1,2,3,4,5,6,7,8,9,11,13)
time.points_2 <- c(1,2,3,4,5,6,7,8,9,10,11,12,13)

time.points <- list(time.points_1, time.points_1, time.points_2, time.points_2)
#time.points <- list(Patient_1_rep_1_filter[1,], Patient_1_rep_2_filter[1,], Patient_2_rep_1_filter[1,], Patient_2_rep_2_filter[1,])

decay_rates = 0.2
```



**Run `dynGENIE3` with default parameters or by tuning the parameters:**

```{r eval=FALSE, include=FALSE}
run_analysis <- dynGENIE3(normalized_counts_list, time.points, alpha = decay_rates, tree.method = "RF", SS.data=NULL, K = "sqrt", ntrees = 1000, regulators = NULL, ncores = 12, verbose = FALSE, seed = NULL)
```



#Filter and visualize the results
First, we plot the distribution of link.list, to help select a threshold.


```{r eval=FALSE, include=FALSE}
link.list <- get.link.list(run_analysis$weight.matrix)
links_filtered_TF <- filter(link.list, link.list$weight >= 0.005) #22383 links

ggplot(link.list, aes(x=weight)) + 
  geom_histogram(color="black", fill="white") +
  labs(title="Links weight distribution", subtitle="Edges weight histogram plot") +
  theme(plot.title = element_text(hjust = 0.5, size = 12, face = "bold"), plot.subtitle = element_text(hjust = 0.5)) +
  xlab("Weight")+ylab("Count")

ggplot(links_filtered_TF, aes(x=weight)) + 
  geom_histogram(color="black", fill="white") +
  xlim(min(links_filtered_TF$weight), max(links_filtered_TF$weight)) +
  labs(title="Links filtered weight distribution", subtitle="Edges weight histogram plot") +
  theme(plot.title = element_text(hjust = 0.5, size = 12, face = "bold"), plot.subtitle = element_text(hjust = 0.5)) +
  xlab("Weight")+ylab("Count") + geom_vline(xintercept= min(links_filtered_TF$weight), size = 2, colour = "red")
```



#Giant component extraction


```{r}
g1 <- graph.data.frame(links_filtered_TF, directed = TRUE) #create graph from datatable
g2 = giant_component_extract(g1, directed = TRUE) #extract the largest connected component of the Graph
giant_comp = g2[[1]]

vcount(g1) #9404
ecount(g1) #22383
vcount(giant_comp) #8620
ecount(giant_comp) #21922

g = igraph::as_data_frame(giant_comp)
names(g)[names(g) == "from"] <- "Regulatory.Gene"
names(g)[names(g) == "to"] <- "Target.Gene"
```



#TFs


```{r}
TFs = read.csv("Data/TF_names_v_1.01.txt", header = FALSE)
names(TFs)[names(TFs) == "V1"] <- "TF_name"

#Number of genes
TF_regulatory_nb = g[(g$Regulatory.Gene %in% TFs$TF_name),]
TF_regulatory_nb[2:3] <- list(NULL)
names(TF_regulatory_nb)[names(TF_regulatory_nb) == "Regulatory.Gene"] <- "Regulatory TF"
TF_regulatory_nb = distinct(TF_regulatory_nb) #Number of regulatory TFs in the network
TF_target_nb = g[(g$Target.Gene %in% TFs$TF_name),]
TF_target_nb[,-c(0,2)] <- list(NULL)
names(TF_target_nb)[names(TF_target_nb) == "Target.Gene"] <- "Target TF"	
TF_target_nb = distinct(TF_target_nb) #Number of target TFs in the network
TF_nb_inter = as.data.frame(intersect(TF_regulatory_nb$`Regulatory TF`, TF_target_nb$`Target TF`))
names(TF_nb_inter)[names(TF_nb_inter) == "intersect(TF_regulatory_nb$`Regulatory TF`, TF_target_nb$`Target TF`)"] <- "Regulatory TF ∩ Target TF"
TF_nb = bind_rows(TF_regulatory_nb,setNames(TF_target_nb,names(TF_regulatory_nb))) %>% unique
names(TF_nb)[names(TF_nb) == "Regulatory TF"] <- "TF name"

non_TF_regulatory_nb = g[!(g$Regulatory.Gene %in% TFs$TF_name),] 
non_TF_regulatory_nb[2:3] <- list(NULL)
names(non_TF_regulatory_nb)[names(non_TF_regulatory_nb) == "Regulatory.Gene"] <- "Regulatory non TF"
non_TF_regulatory_nb = distinct(non_TF_regulatory_nb) #Number of regulatory non TFs in the network (should be equal to 0)
non_TF_target_nb = g[!(g$Target.Gene %in% TFs$TF_name),]
non_TF_target_nb[,-c(0,2)] <- list(NULL)
non_TF_target_nb = distinct(non_TF_target_nb) #Number of target non TFs in the network
names(non_TF_target_nb)[names(non_TF_target_nb) == "Target.Gene"] <- "Target non TF"	
non_TF_nb = bind_rows(non_TF_regulatory_nb,setNames(non_TF_target_nb,names(non_TF_regulatory_nb))) %>% unique
names(non_TF_nb)[names(non_TF_nb) == "Regulatory non TF"] <- "non TF name"
non_TF_nb_inter = as.data.frame(intersect(non_TF_regulatory_nb$`Regulatory non TF`, non_TF_target_nb$`Target non TF`))
names(non_TF_nb_inter)[names(non_TF_nb_inter) == "intersect(non_TF_regulatory_nb$`Regulatory non TF`, non_TF_target_nb$`Target non TF`)"] <- "Regulatory non TF ∩ Target non TF"

#Number of links
non_TF_links = g[!((g$Regulatory.Gene %in% TFs$TF_name) | (g$Target.Gene %in% TFs$TF_name)),] #links number without TFs (should be equal to 0)
TF_links = g[(g$Regulatory.Gene %in% TFs$TF_name) | (g$Target.Gene %in% TFs$TF_name),] #links number containing TF(s)
TF_regulatory = g[(g$Regulatory.Gene %in% TFs$TF_name),]
TF_target = g[(g$Target.Gene %in% TFs$TF_name),]
TFs_inter = g[(g$Regulatory.Gene %in% TFs$TF_name) & (g$Target.Gene %in% TFs$TF_name),] #number of interactions between TFs only
```



#Study of the proportion of TF in the network


```{r}
#Doughnut chart links with vs without TF
df1 = data.frame(value = c(nrow(TF_links), nrow(non_TF_links)), group =c("TF links", "non TF links"))

df1$fraction = df1$value / sum(df1$value)
df1$ymax = cumsum(df1$fraction)
df1$ymin = c(0, head(df1$ymax, n=-1))
df1$labelPosition <- (df1$ymax + df1$ymin) / 2
df1$label <- paste0("count: ", df1$value)

ggplot(df1, aes(ymax=ymax, ymin=ymin, xmax=4, xmin=3, fill=group)) +
  geom_rect() +
  geom_label( x=3.5, aes(y=labelPosition, label=label),
                        show.legend = FALSE) +
  scale_fill_brewer(palette=4) +
  coord_polar(theta="y") +
  xlim(c(2, 4)) +
  scale_fill_brewer(palette="Reds") +
  labs(x="",y="") + 
  theme_minimal() + 
  theme(axis.text = element_blank()) +
  ggtitle(" Proportion of links containing TFs in the network") +
  theme(plot.title = element_text(hjust = 0.5, face = "bold"))

#Pie chart TF vs non TF
df2 = data.frame(value = c(nrow(TF_nb), nrow(non_TF_nb)), group =c("TFs", "non TFs"))

ggplot(df2, aes(x = "", y = value, fill = group)) +
  geom_col() +
  coord_polar(theta = "y") +
  geom_label(aes(label = value), 
             position = position_stack(vjust = 0.5), 
             show.legend = FALSE) +
  scale_fill_brewer(palette="Reds") +
  labs(x="",y="") +
  theme_minimal() +
  ggtitle("TFs proportion in the network") +
  theme(plot.title = element_text(hjust = 0.5, face = "bold"))
  

# Draw pairwise venn diagram
#regulatory TF vs target TF (links)
a = nrow(TF_regulatory)
b = nrow(TF_target)
c = nrow(TFs_inter)
#d = (nrow(TF_regulatory) + nrow(TF_target)) - nrow(TFs_inter)
x <- list(A = 1:(a+c), B = (a+1):(a+b+c))

p = ggVennDiagram(x[1:2], category.names = c("Source TF links", "Target TF links"))
p + labs(title="Links containing TF organization", 
         caption="Total number of links : 3489") +
    theme(plot.title = element_text(hjust = 0.5, size = 14, face = "bold"), 
          plot.caption = element_text(face = "italic", color = "gray40"))

#(non) regulatory TF vs (non) target TF (genes)
a1 = nrow(TF_regulatory_nb)
b1 = nrow(TF_target_nb)
c1 = nrow(TF_nb_inter)
x1 <- list(A = 1:(a1+c1), B = (a1+1):(a1+b1+c1))

a2 = nrow(non_TF_regulatory_nb)
b2 = nrow(non_TF_target_nb)
c2 = nrow(non_TF_nb_inter)
x2 <- list(A = 1:(a2+c2), B = (a2+1):(a2+b2+c2))

p1 = ggVennDiagram(x1[1:2], category.names = c("Source TFs", "Target TFs"))
p1 = p1 + labs(title="TFs organization", 
         caption="Total number of TFs : 718") +
    theme(plot.title = element_text(hjust = 0.5, size = 14, face = "bold"), 
          plot.caption = element_text(face = "italic", color = "gray40"))

p2 = ggVennDiagram(x2[1:2], category.names = c("Source non TFs", "Target non TFs"))
p2 = p2 + labs(title="Non TFs organization", 
         caption="Total number of non TFs : 7902") +
    theme(plot.title = element_text(hjust = 0.5, size = 14, face = "bold"), 
          plot.caption = element_text(face = "italic", color = "gray40"))

grid.arrange(p1, p2, nrow=1, ncol=2)
```

This network contains too many non TF nodes and links, more particularly non TFs target nodes.

This script highlights TF proportion in the network but the high quantity of nodes makes manipulation of the data difficult in order to extract information from it. The script *TF_inferred_network.Rmd* pursues the analysis of the network by using the list of human TFs as regulators, but also as targets (all the nodes of the network correspond to TFs).
