---
title: "GRN Inference from CLL time-course bulk RNA-seq"
author: "Malvina Marku, Flavien Raynal"
output:
  html_document:
    df_print: paged
  html_notebook: default
  pdf_document: default
---

This is an [R Markdown](http://rmarkdown.rstudio.com) Notebook. When you execute code within the notebook, the results appear beneath the code. 

Try executing this chunk by clicking the *Run* button within the chunk or by placing your cursor inside it and pressing *Ctrl+Shift+Enter*. 

**1. Load the libraries and `dynGENIE3` functions:**

```{r}
library(reshape2)
library(doRNG)
library(doParallel)
library(tidyverse)

source("dynGENIE3.R")
```

**2. Load the data**
In our case, we have the time-course transcriptomics from 2 patients and 2 replicates per patient. the inference method will be applied in different ways:
\item patient 1 and 2 taken separately. The average between replicates will be calculated per patient.
\item ...

```{r}
Patient_1_rep_1_table <- read.expr.matrix("/mnt/SERVER-CRCT-STORAGE/CRCT21/Private/CLL_dec2020/RNAseqBulk_Eq21_Eq9/Gene_counts_tables/Gene_counts_Patient1_rep1_TPM.txt",form = "rows.are.genes")

Patient_1_rep_2_table <- read.expr.matrix("/mnt/SERVER-CRCT-STORAGE/CRCT21/Private/CLL_dec2020/RNAseqBulk_Eq21_Eq9/Gene_counts_tables/Gene_counts_Patient1_rep2_TPM.txt", form = "rows.are.genes")

Patient_2_rep_1_table <- read.expr.matrix("/mnt/SERVER-CRCT-STORAGE/CRCT21/Private/CLL_dec2020/RNAseqBulk_Eq21_Eq9/Gene_counts_tables/Gene_counts_Patient2_rep1_TPM.txt", form = "rows.are.genes")

Patient_2_rep_2_table <- read.expr.matrix("/mnt/SERVER-CRCT-STORAGE/CRCT21/Private/CLL_dec2020/RNAseqBulk_Eq21_Eq9/Gene_counts_tables/Gene_counts_Patient2_rep2_TPM.txt", form = "rows.are.genes")
```

**Re-format the data:**

_Firstly_, make sure there is no `NA, NaN, or Inf` value in the dataset: 

```{r}
#Identify the NA, NaN and Inf values
impute.mean <- function(x) replace(x, is.na(x) | is.nan(x) | is.infinite(x), mean(x, na.rm = TRUE))

Patient_1_rep_1_table <- apply(Patient_1_rep_1_table, 2, impute.mean)
sum( apply( Patient_1_rep_1_table, 2, function(.) sum(is.infinite(.))) )

Patient_1_rep_2_table <- apply(Patient_1_rep_2_table, 2, impute.mean)
sum( apply( Patient_1_rep_2_table, 2, function(.) sum(is.infinite(.))) )


```
_Secondly_, re-format the time-series:

```{r}
#Add a row in each of the datasets indicating the time point

Patient_1_time_points <- c(1:ncol(Patient_1_rep_1_table))
Patient_2_time_points <- c(1:ncol(Patient_2_rep_1_table))

Patient_1_rep_1_table <- rbind(Patient_1_time_points, Patient_1_rep_1_table)
Patient_1_rep_2_table <- rbind(Patient_1_time_points, Patient_1_rep_2_table)
Patient_2_rep_1_table <- rbind(Patient_2_time_points, Patient_2_rep_1_table)
Patient_2_rep_2_table <- rbind(Patient_2_time_points, Patient_2_rep_2_table)
# write.table(Patient_1_rep_1_table, file = "Patient_1_rep_1_table.txt")
# write.table(Patient_1_rep_2_table, file = "Patient_1_rep_2_table.txt")
# write.table(Patient_2_rep_1_table, file = "Patient_2_rep_1_table.txt")
# write.table(Patient_2_rep_2_table, file = "Patient_2_rep_2_table.txt")
# 
# Patient_1_rep_1_table <- read.expr.matrix("Patient_1_rep_1_table.txt",form = "rows.are.genes")
# Patient_1_rep_2_table <- read.expr.matrix("Patient_1_rep_2_table.txt", form = "rows.are.genes")
# Patient_2_rep_1_table <- read.expr.matrix("Patient_2_rep_1_table.txt", form = "rows.are.genes")
# Patient_2_rep_2_table <- read.expr.matrix("Patient_2_rep_2_table.txt", form = "rows.are.genes")


time.points <- list(Patient_1_rep_1_table[1,], Patient_1_rep_2_table[1,], Patient_2_rep_1_table[1,], Patient_2_rep_2_table[1,])

time_series_data <- list(Patient_1_rep_1_table[2:nrow(Patient_1_rep_1_table),], Patient_1_rep_2_table[2:nrow(Patient_1_rep_2_table),], Patient_2_rep_1_table[2:nrow(Patient_2_rep_1_table),], Patient_2_rep_2_table[2:nrow(Patient_2_rep_2_table),])
```
**Run `dynGENIE3` with default parameters or by tuning the parameters:**

```{r}
#run_analysis <- dynGENIE3(time_series_data, time.points)
```